registry/storage/tagstore.go

```diff
@@ -9,45 +12,61 @@ import (
 	storagedriver "github.com/docker/distribution/registry/storage/driver"
 )
 
+var _ distribution.TagService = &tagStore{}
+
 // tagStore provides methods to manage manifest tags in a backend storage driver.
+// This implementation uses the same on-disk layout as the (now deleted) signature
+// store.  This provides backward compatibility with current registry deployments
+// which only makes use of the Digest field of the returned distribution.Descriptor
+// but does not enable full roundtripping of Descriptor objects
 type tagStore struct {
 	repository *repository
 	blobStore  *blobStore
-	ctx        context.Context
 }
 
-// tags lists the manifest tags for the specified repository.
-func (ts *tagStore) tags() ([]string, error) {
-	p, err := pathFor(manifestTagPathSpec{
+// Enumerate enumerates the tags stored in this tag store.
+func (ts *tagStore) Enumerate(ctx context.Context, tags []string, last string) (n int, err error) {
+	if len(tags) == 0 {
+		return 0, errors.New("no space in slice")
+	}
+
+	pathSpec, err := pathFor(manifestTagPathSpec{
 		name: ts.repository.Name(),
 	})
-
 	if err != nil {
-		return nil, err
+		return 0, err
 	}
 
-	var tags []string
-	entries, err := ts.blobStore.driver.List(ts.ctx, p)
+	entries, err := ts.blobStore.driver.List(ctx, pathSpec)
 	if err != nil {
 		switch err := err.(type) {
 		case storagedriver.PathNotFoundError:
-			return nil, distribution.ErrRepositoryUnknown{Name: ts.repository.Name()}
+			return 0, distribution.ErrRepositoryUnknown{Name: ts.repository.Name()}
 		default:
-			return nil, err
+			return 0, err
 		}
 	}
 
+	sort.Strings(entries)
+	var tagNames []string
 	for _, entry := range entries {
-		_, filename := path.Split(entry)
+		if len(tagNames) == len(tags) {
+			copy(tags, tagNames)
+			return len(tags), nil
+		}
 
-		tags = append(tags, filename)
+		_, filename := path.Split(entry)
+		if filename > last {
+			tagNames = append(tagNames, filename)
+		}
 	}
 
-	return tags, nil
+	n = copy(tags, tagNames)
+	return n, io.EOF
```

Comment:
@aaronlehmann A more efficient, cached version should be possible. We just have to commit to the possibility this interface may change.